core/shared/src/main/scala/org/typelevel/log4cats/LoggerKernel.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2018 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.log4cats
+
+/**
+ * This is the fundamental abstraction: a single-abstract-method interface that has the following
+ * properties:
+ *
+ *   - Doesn't enforce a specific memory layout. More specific interfaces can back `LogRecord` with
+ *     mutables/immutables values, and avoid storing things that are not important.
+ *   - LogRecord allows to precisely capture a lot of information. In particular, it does not
+ *     enforce a `Map[String, String]` representation of context values that is not sufficient to
+ *     leverage all the power from logging backends query engines, and without pulling a third-party
+ *     JSON library.
+ *   - The SAM-like nature of the construct makes it inherently middleware friendly, as a single
+ *     method call needs to be intercepted/proxied in order to amend the behaviour of the logger.
+ *
+ * This also means that different libraries can use wrappers on top of this kernel interface to use
+ * whatever UX is preferred without necessarily imposing constraints on the underlying
+ * implementation.
+ */
+trait LoggerKernel[F[_], Ctx] {
+  type Builder = Log.Builder[Ctx]
+
+  def log(level: KernelLogLevel, record: Builder => Builder): F[Unit]
+
+  final def logTrace(record: Builder => Builder): F[Unit] =
+    log(KernelLogLevel.Trace, record)
+  final def logDebug(record: Builder => Builder): F[Unit] =
+    log(KernelLogLevel.Debug, record)
+  final def logInfo(record: Builder => Builder): F[Unit] =
+    log(KernelLogLevel.Info, record)
+  final def logWarn(record: Builder => Builder): F[Unit] =
+    log(KernelLogLevel.Warn, record)
+  final def logError(record: Builder => Builder): F[Unit] =
+    log(KernelLogLevel.Error, record)
```

Comment:
Here's what the inlined version looks like:

```suggestion
  def log(level: KernelLogLevel, record: Log.Builder[Ctx] => Log.Builder[Ctx]): F[Unit]

  final def logTrace(record: Log.Builder[Ctx] => Log.Builder[Ctx]): F[Unit] =
    log(KernelLogLevel.Trace, record)
  final def logDebug(record: Log.Builder[Ctx] => Log.Builder[Ctx]): F[Unit] =
    log(KernelLogLevel.Debug, record)
  final def logInfo(record: Log.Builder[Ctx] => Log.Builder[Ctx]): F[Unit] =
    log(KernelLogLevel.Info, record)
  final def logWarn(record: Log.Builder[Ctx] => Log.Builder[Ctx]): F[Unit] =
    log(KernelLogLevel.Warn, record)
  final def logError(record: Log.Builder[Ctx] => Log.Builder[Ctx]): F[Unit] =
    log(KernelLogLevel.Error, record)
```
The readability is about the same, and it means that `NoOpKernelLogger` can be implemented like this:
```scala
object NoOpLoggerKernel {
  def apply[F[_]: Applicative, Ctx]: LoggerKernel[F, Ctx] = (level, record) => Applicative[F].unit
}
```